extension/src/topn.rs

```diff
@@ -0,0 +1,325 @@
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Internal<InternalTopN>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            match (state1, state2) {
+                (None, None) => None,
+                (None, Some(state2)) => Some(state2.clone().into()),
+                (Some(state1), None) => Some(state1.clone().into()),
+                (Some(state1), Some(state2)) => Some(
+                    InternalTopN::combine(&state1, &state2).into())
+            }
+        })
+    }
+}
+
+// PG function to generate a user-facing TopN object from a InternalTopN.
+#[pg_extern(immutable, parallel_safe, schema = "toolkit_experimental")]
+fn topn_final(
+    state: Option<Internal<InternalTopN>>,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<toolkit_experimental::TopN<'static>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            let state = match state {
+                None => return None,
+                Some(state) => state,
+            };
+
+            TopN::from_internal_topn(&state).into()
+        })
+    }
+}
+
+extension_sql!(r#"
+CREATE AGGREGATE toolkit_experimental.topn_agg(size int, value int)
+(
+    sfunc = toolkit_experimental.topn_trans,
+    stype = internal,
+    finalfunc = toolkit_experimental.topn_final,
+    combinefunc = toolkit_experimental.topn_combine,
+    serialfunc = toolkit_experimental.topn_serialize,
+    deserialfunc = toolkit_experimental.topn_deserialize,
+    parallel = safe
+);
+"#);
+
+#[pg_extern(immutable, parallel_safe, schema = "toolkit_experimental")]
+pub fn topn_compound_trans(
+    state: Option<Internal<InternalTopN>>,
+    value: Option<toolkit_experimental::TopN<'static>>,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Internal<InternalTopN>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            match (state, value) {
+                (a, None) => a,
+                (None, Some(a)) => Some(a.to_internal_topn().into()),
+                (Some(a), Some(b)) =>
+                    Some(InternalTopN::combine(&a, &b.to_internal_topn()).into()),
+            }
+        })
+    }
+}
+
+extension_sql!(r#"
+CREATE AGGREGATE toolkit_experimental.rollup(
+    toolkit_experimental.topn
+) (
+    sfunc = toolkit_experimental.topn_compound_trans,
+    stype = internal,
+    finalfunc = toolkit_experimental.topn_final,
+    combinefunc = toolkit_experimental.topn_combine,
+    serialfunc = toolkit_experimental.topn_serialize,
+    deserialfunc = toolkit_experimental.topn_deserialize,
+    parallel = safe
+);
+"#);
+
+//---- Available PG operations on the topn structure
+#[pg_extern(immutable, parallel_safe, schema = "toolkit_experimental")]
+pub fn num_vals(
+    agg: toolkit_experimental::TopN<'static>,
+) -> int {
+    agg.total_inputs as _
+}
+
+#[pg_extern(immutable, parallel_safe, name="topn", schema = "toolkit_experimental")]
+pub fn topn_iter (
+    n: i32,
+    agg: toolkit_experimental::TopN<'static>,
```

Comment:
```suggestion
    agg: toolkit_experimental::TopN<'input>,
```

This should _never_ be `'static` or the same as the output; that claims that the input argument will live forever when it only lives for this one call.

Reply:
This actually couldn't be made non-static without breaking the compilation.  After discussing this we've come up with an explanation of why this is okay, which has been added as a comment here.